src/common/WrrQueue.h

```diff
@@ -0,0 +1,344 @@
+    }
+    total_priority += priority;
+    SubQueue *sq = &queue[priority];
+    return sq;
+  }
+
+public:
+  WrrQueue(unsigned max_per, unsigned min_c)
+    : total_priority(0),
+      high_size(0),
+      wrr_size(0),
+      max_cost(0),
+      dq(queue.rbegin())
+  {
+    srand(time(0));
+  }
+
+  unsigned length() const final {
+    return high_size + wrr_size;
+  }
+
+  void remove_by_filter(std::function<bool (T)> f, std::list<T> *removed = 0) final {
+    for (typename SubQueues::iterator i = queue.begin();
+	 i != queue.end();
+	 ) {
+      wrr_size -= i->second.remove_by_filter(f, removed);
+      ++i;
+    }
+    for (typename SubQueues::iterator i = high_queue.begin();
+	 i != high_queue.end();
+	 ) {
+      high_size -= i->second.remove_by_filter(f, removed);
+      if (i->second.empty()) {
+	high_queue.erase(i++);
+      } else {
+	++i;
+      }
+    }
+  }
+
+  void remove_by_class(K k, std::list<T> *out = 0) final {
+    for (typename SubQueues::iterator i = queue.begin();
+	 i != queue.end();
+	 ) {
+      wrr_size -= i->second.remove_by_class(k, out);
+      ++i;
+    }
+    for (typename SubQueues::iterator i = high_queue.begin();
+	 i != high_queue.end();
+	 ) {
+      high_size -= i->second.remove_by_class(k, out);
+      if (i->second.empty()) {
+	high_queue.erase(i++);
+      } else {
+	++i;
+      }
+    }
+  }
+
+  void enqueue (K cl, T item, unsigned priority, unsigned cost = 0,
+		bool front = CEPH_OP_QUEUE_BACK,
+		unsigned opclass = CEPH_OP_CLASS_NORMAL) final {
+    switch (opclass){
+      case CEPH_OP_CLASS_NORMAL :
+	if (cost > max_cost) {
+	  max_cost = cost;
+	}
+	create_queue(priority)->enqueue(cl, cost, item, front);
+	++wrr_size;
+	break;
+      case CEPH_OP_CLASS_STRICT :
+	high_queue[priority].enqueue(cl, 0, item, front);
+	++high_size;
+	break;
+      default :
+	assert(1);
+    }
+  }
+
+  bool empty() const final {
+    assert(total_priority >= 0);
+    assert((total_priority == 0) || !(queue.empty()));
+    return (high_size + wrr_size  == 0) ? true : false;
+  }
+
+  void inc_dq() {
+    ++dq;
+    if (dq == queue.rend()) {
+      dq = queue.rbegin();
+    }
+  }
+
+  T dequeue() final {
+    assert(!empty());
+
+    if (!(high_queue.empty())) {
+      T ret = high_queue.rbegin()->second.front().second;
+      high_queue.rbegin()->second.pop_front();
+      if (high_queue.rbegin()->second.empty()) {
+	high_queue.erase(high_queue.rbegin()->first);
+      }
+      --high_size;
+      return ret;
+    }
+    while (true) {
+      if (dq->second.size() != 0) {
+	if (dq->second.size() == wrr_size) {
+	  break;
+	}
+	else {
+	  // flip coin to see if the previous prioirty gets to run again.
+	  if ((rand() % total_priority) <= dq->first) {
+	    if (dq->second.front().first == 0) {
+	      break;
+	    }
+	    else {
+	      // flip a coin to see if this priority gets to run based on cost
+	      if ((rand() % max_cost) <=
+		  (max_cost - (unsigned)((double)dq->second.front().first * 0.9))){
```

Comment:
I don't understand why this works (not that it doesn't, it's just not obvious to me).  It would be really helpful if you could add a detailed comment here explaining why this gives you the right properties.